File: 백준/Bronze/2231. 분해합/분해합.py
def splitsum(n):
    g = str(n)
    leng = len(g)
    sum = n
    for i in g:
        sum += int(i)
    return sum

n = input()
leng = len(n)
n = int(n)
no = True
for i in range(n-leng*9, n):
    if i < 1:
        continue
    if splitsum(i) == n:
        print(i)
        no = False
        break
if no:
    print(0)

File: 백준/Bronze/2231. 분해합/분해합.cpp
#include <iostream>
#include <cmath>
using namespace std;

int main() {
	int n, len;
	cin >> n;
	len = log10(n) + 1;
	bool no = true;
	for (int i = max(1, n - len * 9); i < n; i++) {
		int temp = i, sum = i, leng = log10(i);
		int g = pow(10, leng);
		while (g > 0) {
			sum += temp / g;
			temp -= temp / g * g;
			g /= 10;
		}
		if (sum == n) {
			cout << i;
			no = false;
			break;
		}
	}
	if (no)	cout << 0;
}